azurerm/resource_arm_hd_insight_clusters.go

```diff
@@ -0,0 +1,918 @@
+					Schema: map[string]*schema.Schema{
+						"blob_endpoint": {
+							Required: true,
+							ForceNew: true,
+							Type:     schema.TypeString,
+						},
+						"container": {
+							Required: true,
+							ForceNew: true,
+							Type:     schema.TypeString,
+						},
+						"access_key": {
+							Required: true,
+							ForceNew: true,
+							Type:     schema.TypeString,
+						},
+					},
+				},
+			},
+			"tags": tagsSchema(),
+			"tier": {
+				Optional: true,
+				ForceNew: true,
+				Type:     schema.TypeString,
+				Default:  "standard",
+				ValidateFunc: validation.StringInSlice([]string{
+					"standard",
+					"premium",
+				}, true),
+				DiffSuppressFunc: ignoreCaseDiffSuppressFunc,
+			},
+		},
+	}
+}
+
+func resourceArmHDInsightClustersCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).hdInsightClustersClient
+	ctx := meta.(*ArmClient).StopContext
+
+	resourceGroupName := d.Get("resource_group_name").(string)
+	clusterName := d.Get("name").(string)
+	parameters := hdinsight.ClusterCreateParametersExtended{}
+	parameters.Location = utils.String(azureRMNormalizeLocation(d.Get("location").(string)))
+	tags := d.Get("tags").(map[string]interface{})
+	tmpParamOfTags := expandTags(tags)
+	parameters.Tags = tmpParamOfTags
+	parameters.Properties = &hdinsight.ClusterCreateProperties{}
+	if paramValue, paramExists := d.GetOk("cluster_version"); paramExists {
+		parameters.Properties.ClusterVersion = utils.String(paramValue.(string))
+	}
+	if paramValue, paramExists := d.GetOk("os_type"); paramExists {
+		parameters.Properties.OsType = hdinsight.OSType(paramValue.(string))
+	}
+	if paramValue, paramExists := d.GetOk("tier"); paramExists {
+		parameters.Properties.Tier = hdinsight.Tier(paramValue.(string))
+	}
+	parameters.Properties.ClusterDefinition = &hdinsight.ClusterDefinition{}
+	if paramValue, paramExists := d.GetOk("cluster_type"); paramExists {
+		parameters.Properties.ClusterDefinition.Kind = utils.String(paramValue.(string))
+	}
+	if paramValue, paramExists := d.GetOk("component_version"); paramExists {
+		tmpParamOfComponentVersion := make(map[string]*string)
+		for tmpParamKeyOfComponentVersion, tmpParamItemOfComponentVersion := range paramValue.(map[string]interface{}) {
+			parametersPropertiesClusterDefinitionComponentVersion := utils.String(tmpParamItemOfComponentVersion.(string))
+			tmpParamOfComponentVersion[tmpParamKeyOfComponentVersion] = parametersPropertiesClusterDefinitionComponentVersion
+		}
+		parameters.Properties.ClusterDefinition.ComponentVersion = tmpParamOfComponentVersion
+	}
+	tmpParamOfConfigurations := make(map[string]interface{})
+	tmpParamOfGatewayConfigurations := make(map[string]interface{})
+	tmpParamOfGatewayConfigurations["restAuthCredential.isEnabled"] = true
+	tmpParamOfGatewayConfigurations["restAuthCredential.username"] = d.Get("restauth_username")
+	tmpParamOfGatewayConfigurations["restAuthCredential.password"] = d.Get("restauth_password")
```

Comment:
can we parse this out in a separate expand and flatten function? Redis has good examples of this with the configuration

Reply:
This is base on an auto-generation tool which handles all the set and get method. And I don't see any benefits to extract to a lot of local methods just to flatten or expand them. It is hard to see the overall structure.

Reply:
Consistency with every other Provider - it also has the benefit of grouping this such that it's easier to understand

Reply:
I don't think "keep with every other provider" persuade me:
1. The other provider code doesn't look good, it is hard to locate the figure out the structure in such a spaghetti-like code structure.
2. We will never make any steps forward if we always keep consistency with the legacy code.
3. Abstract to functions means reusing, and obviously it is not the case in other providers.

Reply:
> 1. The other provider code doesn't look good, it is hard to locate the figure out the structure in such a spaghetti-like code structure.

This is just an opinion and also not persuasive, probably many people on the team and a number of community contributors would beg to differ.

> 2. We will never make any steps forward if we always keep consistency with the legacy code.

What step forward are you making here? Code generation? Because other providers with code generation adhere to the standards, so this just seems to be a deficiency in your generation.

> 3. Abstract to functions means reusing, and obviously it is not the case in other providers.

Not just reusing, it offers the option of [unit testing](https://github.com/terraform-providers/terraform-provider-azurerm/blob/150e426d6d4c7ede6319b38761de5166fb3708ac/azurerm/tags_test.go#L71-L96), it also is good organization when you have multiple concerns.

An extremely large, complex, function is harder to digest for a human that a smaller one. Breaking it in to smaller, purpose built functions, is more clear. This function has multiple related but different things it is doing:

* Reading state and creating the SDK types
* Calling the SDK and handling the result

Calling the SDK and handling the SDK result seem like a reasonable thing for this function to do, converting resource state to an SDK type using all the associated type assertions is an unnecessary addition complicating the SDK interaction logic with a bunch of unrelated lines that the next person in this code will have to read through until they can figure out where the real action is starting. By separating out to the flatten/expand model, interaction and struct assignment is easily separated mentally.

Reply:
A lot of this comes down to the question of is this solely for auto-generated maintenance or do you intend to fix bugs or flow idiosyncrasies manually in these files?